src_lockfree/bounded_stack.mli

```diff
@@ -0,0 +1,113 @@
+(** Lock-free bounded stack. *)
+
+(** {1 API} *)
+
+type 'a t
+(** Represents a lock-free bounded stack holding elements of type ['a]. *)
+
+val create : ?capacity:int -> unit -> 'a t
+(** [create ~capacity ()] creates a new empty bounded stack with a maximum 
+capacity of [capacity]. Default [capacity] value is [Int.max_int].
+*)
+
+val length : 'a t -> int
+(** [length stack] returns the number of elements currently in the [stack]. *)
+
+val is_empty : 'a t -> bool
+(** [is_empty stack] returns [true] if the [stack] is empty, otherwise [false]. *)
+
+(** {2 Consumer functions} *)
+val peek : 'a t -> 'a
+(** [peek stack] returns the top element of the [stack] without removing it. If 
+the stack is empty, the domain will be suspended until another domain successfully 
+push another element in the stack with [push] or [push_opt] . *)
+
+val peek_opt : 'a t -> 'a option
+(** [peek_opt stack] returns [Some] of the top element of the [stack] without
+    removing it, or [None] if the stack is empty. *)
+
+val pop : 'a t -> 'a
+(** [pop stack] removes and returns the top element of the [stack].
+    Raises an exception if the stack is empty. *)
+
+val pop_opt : 'a t -> 'a option
+(** [pop_opt stack] removes and returns [Some] of the top element of the [stack],
+    or [None] if the stack is empty. *)
+
+val pop_all : 'a t -> 'a list
+(** [pop_all stack] removes and returns all elements of the [stack] in the reverse 
+order they were pushed. *)
+
+(** {2 Producer functions} *)
+
+val push : 'a t -> 'a -> unit
+(** [push stack element] adds [element] to the top of the [stack].
+    Raises an exception if the stack is full. *)
```

Comment:
Looking at the implementation this seems to block rather than raise an exception in case the stack is full.